core/lib/spree/core/search/product.rb

```diff
@@ -0,0 +1,81 @@
+module Spree
+  module Core
+    module Search
+      class Product < Base
+        def initialize(params)
+          super
+          prepare(params)
+        end
+
+        def search
+          @products = get_base_scope
+
+          unless Spree::Config.show_products_without_price
+            @products = @products.merge(Spree::Price.amount_present).merge(Spree::Price.with_currency(current_currency))
+          end
+
+          @products = @products.page(@params[:page]).per(@params[:per_page])
+        end
+
+        protected
+
+        def add_eagerload_scopes(scope)
+          # TL;DR Switch from `preload` to `includes` as soon as Rails starts honoring
+          # `order` clauses on `has_many` associations when a `where` constraint
+          # affecting a joined table is present (see
+          # https://github.com/rails/rails/issues/6769).
```

Comment:
I was unable to reproduce the issue using AR 4.2.3. I'll double-check tomorrow morning and remove the preload after I'm sure it's fixed.